### ISSUE TITLE
Ship adaptive agents end to end: durable execution, evidence acquisition, measured improvement, product adoption

### ISSUE THREAD

**Issue by drewstone** [tangle-network/agent-runtime#1306]
## Outcome
One agent can pursue real work, obtain better evidence, change its approach, retain useful capabilities, and improve later work. The same exact execution path serves Discovery, software-factory work, product requests, and evaluation. Agents may revise tools, skills, code, context acquisition, and organization—not only prompts. Weight training is optional; GEPA/Omni, DSPy/RLM and native harnesses remain upstream implementations.

**Release proof:** one real Discovery pursuit and one Platform-owned product agent complete work, accept an external correction, retain a meaningful non-prompt artifact, evaluate an improvement on fresh work, activate the exact assessed revision, and restore the previous revision. A working mechanism and a measured learning gain are separate claims. Do not manufacture a win to close this issue.

## Execution order / ownership
Start A+B+D at their existing owners; C makes those capabilities product-usable. E is one bounded capability within that path, not a new workstream that delays it. F measures methods while G qualifies actual releases. One active delivery PR per affected repo; extend compatible open work, preserve peer-owned changes, never open empty placeholder PRs. Link commits, tests and deployed receipts under each item. Do not mark source-merged work deployed.

### A — Control and recover actual running agents
**Owner:** Runtime + Sandbox/provider. Existing tracker: #1165; preserve #1291's listener-lifetime repair.
- [ ] Extend existing durable steer/answer/cancel records to the root and native-harness managers. Reuse acknowledgers; preserve request identity across reconnect. Distinguish recorded, delivered, consumed where observable, and settled; report unsupported in-turn delivery explicitly.
- [ ] Preserve completed child results, partial artifacts, pending control requests and uncertain external effects across coordinator interruption. Reconcile existing execution; do not silently spawn replacement work. Make queued settlements/capacity pressure visible to the agent; no research-specific scheduler.
- [ ] Prove through real HTTP/provider boundaries: client disconnect, coordinator restart, late child result, repeated control delivery, cancellation and final cleanup. Keep authorized work moving; qualification must not become another blanket no-research gate.
**Done:** a fresh authenticated client steers/answers a live root and retrieves the original work after interruption; no duplicate effect or falsely confirmed cleanup.

### B — Preserve the exact executable agent and retained state
**Owner:** Runtime candidate/workspace APIs + profile materializer + Sandbox. Reuse `harvestSurfaceDiffs`, candidate bundles, native session/file APIs and Knowledge stores.
- [ ] Carry supported tool definitions AND handlers, skills/helper bytes, knowledge references and native state from capture through evaluation to next execution. A digest-only diff is not retained content; capture before disposal.
- [ ] Keep independent runs/tenants isolated; reset actually removes learned state, carry preserves only declared state. Distinguish same-session continuation, fresh-session import, resume and fork. Reject unsupported profile axes before spend rather than silently ignoring them.
- [ ] Validate every delivered revision against effective materialization, not only profile JSON. Preserve permissions and tool authority; candidate code cannot alter independent assessment or grant itself access.
**Done:** a tool/helper change—not just prompt text—is used by the next real worker; byte/state mismatch, missing handler and cross-tenant reuse fail at the owning boundary.

### C — Connect measured improvement to the existing product
**Owner:** Intelligence/Platform (`agent-dev-container`) + `agent-app`, using Runtime Intelligence APIs.
- [ ] Connect the existing `runAndStorePlatformAgentProfileImprovement` job to an authorized durable product dispatch. Supply the actual product executor, task source and evaluator; reuse existing run/proposal/review/activation records. Recheck callers before adding a producer: prior search found test callers, not proof of a deployed job.
- [ ] Make ordinary execution and evaluation use the same source revision and materializer. Apply the exact approved revision on subsequent work; handle concurrent source edits and idempotent activation/restore through the existing Platform store.
- [ ] Consolidate Agent App's duplicate certified-delivery refresh/cache into Runtime's shared source, preserving force refresh, coalescing and current-profile inspection. Document prompt projection truthfully; do not pretend exposing `current()` implements tool/file delivery.
- [ ] Expose request/status/correction, candidate diff, evidence, cost completeness, activate and restore through existing product surfaces. Keep secrets/authorization server-owned. No second chat shell or activation database.
**Done:** a customer starts an improvement, reviews it, activates it, observes the next worker use it, and restores the previous behavior through the supported interface.

### D — One honest resource account from admission to result
**Owner:** Runtime #1175/#1252 and existing provider grants; Intelligence `auto-improve` adapters.
- [ ] Remove missing-cost-to-zero conversions, including `runFixEngine`'s `costUsd ?? 0`; carry observed, estimated, unreported and explicitly non-token-priced work distinctly. Include failed/losing work, retrieval, analysts, search and final assessment without double-counting.
- [ ] Use existing pre-dispatch reservations/protected grants for enforced limits. Post-iteration checks are reporting, not a hard spending ceiling. Preserve settlement through cancellation and response loss.
- [ ] Give comparison arms explicit resource allowances enforced at the serving path; report actual usage and unused allowances. Fixed N is not equal dollars. Infrastructure and subscription costs remain named separately.
**Done:** missing receipts cannot produce a complete total; actual joined receipts reconcile across failed, cancelled and repeated cumulative events.

### E — Let agents acquire and select better evidence
**Owner:** Knowledge acquisition/retrieval hooks + Runtime dispatch; Eval measures utility. Reuse `runRagKnowledgeImprovementLoop.acquireKnowledge`, research-loop/readiness hooks, existing delegate/spawn, connectors and native RLMs.
- [ ] Connect a concrete information gap to caller-permitted acquisition before or during work: existing retrieval, web/source inspection, a delegated agent, RLM exploration, or a targeted experiment. Invocation is agent/policy-selected, not a mandatory preflight or an always-on classifier.
- [ ] Retain source/version/time/content references, access/usage rights, selection rationale and measured acquisition cost in existing evidence records. Treat fetched content as untrusted data; prevent credential leakage, cross-tenant retention and final-answer contamination. Record what evidence the worker actually consumes.
- [ ] Optimize acquisition/selection as an executable candidate through existing GEPA/Omni/Knowledge surfaces. Compare against existing retrieval and an appropriate resource-matched alternative on downstream accepted work—not document count or an agent's quality rating.
**Done:** a real missing/stale-evidence task triggers acquisition, the next action consumes the result, and an acquisition-disabled control isolates its effect. Optional absence does not block unrelated work.
**Prior art:** [WeCo AutoData, 17 Sep 2026](https://arxiv.org/abs/2609.19754) searches executable document selectors using proxy-model training feedback; [code](https://github.com/WecoAI/AutoData). That is pre-training corpus selection, not a ready-made live web-search agent. Transfer the testable idea—optimize the data-selection procedure—not its training stack. Code is MIT; the published feature bank is CC BY-NC 4.0 and corpus terms are separate. No commercial-data import assumed.

### F — Improve methods and checks, not just prompts
**Owner:** Runtime/Eval #1297/#1299; existing Supervisor Lab #112/#113; Intelligence analyst-quality #2889.
- [ ] Exercise existing full-profile/program, scoped, sequential and joint search against the same real worker. Native harnesses own inner loops. Use one pinned GEPA/Omni route and one appropriate real RLM integration; no DSPy rewrite of working coding harnesses, no local fallback algorithm.
- [ ] Feed actionable execution evidence to candidates. Evaluate proposed working checks against independent known-outcome controls with Eval's existing evaluator audit; do not let self-authored checks certify their own claims.
- [ ] Compare reset, resource-matched retry/search, carry and revise on predeclared fresh source units. Isolate state, preserve failures/missingness and keep outer final evidence outside adaptation. Assess specialist improvement separately from learning-procedure transfer.
**Done:** retain executable candidates and a reproducible result—positive, negative or inconclusive—with complete scope and costs. Extend the current Lab instrument; do not copy it into Discovery.

### G — Ship one compatible path to two consumers
**Owner:** release owners + Discovery Lab + the existing Platform/Agent App product.
- [ ] Publish/adopt an exact compatible package set, rebundle frozen Discovery methods, and verify served revisions. Do not widen peer ranges or mutate historical run registrations to claim compatibility.
- [ ] Run the release proof above in real Discovery work and a product workflow. Preserve external verification, effective state, control acknowledgements, usage, artifacts and restore evidence. One documented clean-checkout command per consumer; no machine-local patches.
- [ ] Keep Discovery research goals/methods agent-authored; its docs own doctrine, Lab owns runs/evidence, lower packages own mechanisms. Carry useful evidence across pursuits without adding a Discovery execution schema or fixed swarm topology.
**Done:** another engineer reproduces both journeys; observed defects are fixed upstream and covered by regression tests. Synthetic fixtures establish contracts, not native-model capability or a learning gain.

## Build discipline
Small cohesive owner modules; no omnibus agent/controller class. A new abstraction must remove an existing implementation or solve a demonstrated consumer gap. Keep public API surface minimal, preserve useful lower-level composition, and remove replaced paths in the same delivery. Failure-first regressions, cancellation/concurrency/security coverage and packed cross-package consumer tests are part of each change—not separate fluff tasks. No protection bypass, speculative algorithm implementation, forced multi-agent topology, new ledger/cache/bus, or unrelated version sweep.

## Current baseline / delivery links
Runtime training/provenance and consolidation #1287/#1300/#1302/#1304 are merged; they do not establish this product. Supervisor Lab [#112](https://github.com/tangle-network/supervisor-lab/pull/112) connects experiments; [#113](https://github.com/tangle-network/supervisor-lab/issues/113) retains their live acceptance work. This issue coordinates shared delivery and references those owners rather than replacing their histories.

Implementation PRs and exact acceptance evidence will be linked here as delivered. None of the unchecked items is claimed complete by opening this issue.

**Comment by drewstone** [comment]
## Delivery: #1311 — durable root corrections and parent escalation

Open, non-draft, based on current main; preserves merged #1308/#1310/#1307.

- Root corrections now use the existing `writeWorkerSteer` request/claim/acknowledgement protocol. Router roots read between turns; native roots use their accepting inbox or report unsupported. Same-operation retries do not redeliver; prior unknown claims remain unknown; unseen final requests expire without delivery.
- Fixed `supervise()` silently dropping its configured `escalateQuestion` handler for both root and nested managers. The application's inbox retains persistence and authorization ownership.
- No new public export, controller, ledger, control schema, MCP authority or dependency.

Exact-head clean-runner verification: **4,324 tests pass / six existing skips**, lint, source/example types, packed consumers, edge execution and generated docs. Nine new tests produce eight expected failures plus one passing existing-behavior control against the original implementation. Full source and logs: https://github.com/tangle-network/agent-runtime/actions/runs/35544779008 . A reproduced startup-readiness test assumption was repaired with its existing start signals; control-response deadlines remain unchanged and disabling actual delivery still fails the test.

**A remains partial:** durable root answers, a file-backed question inbox, and real provider/product qualification are not delivered by this PR. Inbox acceptance is not model consumption. No merge, release or deployment performed. Normal PR CI and independent review are tracked on #1311.

Next highest-value product work remains dispatching the existing improvement job through the actual product executor, then proving exact activation and restore. Acquisition stays a bounded part of that workflow, not a new training project.

**Comment by x711criptic** [comment]
If this is about agent tooling — x711 has 26 pay-per-use tools (web search, prices, shared memory, tx simulation) with a free tier. No signup to start, MCP supported.

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"price_feed","query":"ETH"}'
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by drewstone** [comment]
Discovery evidence for B is now merged in tangle-network/discovery-lab#893 at c7b178168c41cc57677a6c0b4c708c49fed967c6.

The completed September 20 program used a registration-local capture wrapper, now explicitly retired. Its automatic assessor capture exceeded the Sandbox 65,536-character argument bound and refused deletion. The wrapper can also forward an already-aborted research signal to provider destruction; Runtime environment-provider.ts itself calls destroy() without that signal. The executed method bytes and both defects remain retained.

The operator separately recovered and hash-verified all 59 selected assessment files (392,572 bytes). All 15 scientific environments were checked stopped or absent, and driver sandbox-6a5126879216 was stopped at 23:24Z, before the original deadline. This is operational recovery, not proof of a generic atomic checkpoint.

Source evidence: https://github.com/tangle-network/discovery-lab/tree/c7b178168c41cc57677a6c0b4c708c49fed967c6/prereg/original-science-20260920

Shared retention remains open under B. Existing Runtime candidate/workspace and provider managed checkpoint/fork APIs are the owning surfaces. A supported contract still needs capture-before-disposal, a durable receipt, independently bounded cleanup, and demonstrated restoration. No new Lab capture wrapper or scheduler should replace them. #1312's bounded worker-output retrieval is a distinct improvement; it does not settle B or prove an autonomous research/check/revision cycle.

**Comment by drewstone** [comment]
Runtime #1312 is merged at `3789e6933fa1873b12d4f242004f9fea0e925fe2`, preparing 0.248.0. It consolidates usage accounting, preserves unknown receipts and separate estimates across incremental reports, leaves durable root steers pending until the harness inbox is ready, and adds generic bounded retained-output reads.

All four final CI jobs passed, including 4,361 tests with six skips, packed cohort, Bench, and official optimizer checks. Independent Luna reviews and final GitHub review completed. A 40,588,123-byte retained Discovery artifact remained complete; full traversal used 2,468 pages with one read and one serialization in a local adapter measurement.

No package was published by this task. The coordinated release owner is publishing Knowledge 17.1.0; a follow-up Runtime compatibility lane will verify Eval 0.183 with that published cohort. Durable root answers, executable-state capture before disposal, downstream adoption, end-to-end qualification, and demonstrated learning gains remain open. No umbrella checkbox is marked complete by this source delivery.

**Comment by drewstone** [comment]
Runtime #1314 is merged at `1f3b5bfa5d661f4e812bc80f8444622033b5f46f`, preparing Runtime 0.249.0 and Bench 0.13.7.

It qualifies the published Eval 0.183.0 and Knowledge 17.1.0 cohort without Runtime implementation changes. This Runtime release requires Eval `>=0.183.0 <0.184.0`, matching its tested development minor. Runtime and Bench share one installed Eval and Knowledge copy.

All four CI jobs passed: 4,361 tests with six skips, exact packed cohort, Bench, and official optimizers. Independent Luna reviews and the final GitHub automated review passed. The merged tree matches the reviewed branch and tested CI merge tree.

Publication and downstream App adoption were handed to the coordinated release owner with the exact source SHA and maintained workflow. No package was published by this task. Durable root answers, executable-state capture before disposal, consumer deployment, end-to-end qualification, and demonstrated learning gains remain open under this umbrella.
